### ISSUE TITLE
Training time and parameters

### ISSUE THREAD

**Issue by darrenjkt** [China-LiuXiaopeng/BraTS-DMFNet#1]
Hi, you mentioned that your model took under 10 hours to train on 4 gtxforce 1080Ti GPUs. 

I'm running mostly the same configuration as DMFNet_GDL_all.yaml
- 267 training patient data
- DataLoader workers = 8
- batch_size = 8
- 4 P100 GPUs (which are faster and higher memory than 1080Ti)

However, without altering your code, each epoch takes me around 15minutes, which means that I'll be taking 125hrs to finish training 500epochs. 

What other setups did you have? Or do you know what might be wrong on my end? 

Thank you!

**Comment by China-LiuXiaopeng** [comment]
Hi. (1) I have checked the config and it seems not problems, no other setups are needed. (2) It runs on  four 1080Ti just need 10 hours, I ever tried on four 2080Ti gpus which run at a significantly faster speed! I believe that P100s should achieve the same performance of the running time though I never try. (3) Could you provide more information about the time bottleneck during training, i.e., data loading? network forward computing? loss or gradient computing? It would be helpful! Please let me know if anyone face the similar issue?

**Comment by darrenjkt** [comment]
Hi thank you for getting back to me. 

My main time consumption seems to be from the data loading. Once it finishes the dataloading part, the rest are pretty fast. I've played around with different amounts of data workers, 3 seems to work best for me. When I try 8, it often takes a really long time. 

I will run it again on my side, time the different subprocess and I'll let you know. 

I forgot to mention, I'm running your model on BraTS2019 training data though I don't think that's too big of a difference? 

Also, did your training loss look like this? It keeps fluctuating between 0.25 and 0.4 for me.
<img width="753" alt="Screen Shot 2019-08-22 at 7 47 37 am" src="https://user-images.githubusercontent.com/39115809/63632410-acdc3a80-c678-11e9-95e4-8c6619358fb5.png">

**Comment by darrenjkt** [comment]
You mentioned that training over 288 patients, batchsize=8 for 500 epochs would be 10hrs, which means that one epoch (36 batches) would take around 1 min 12s per epoch. Some batches alone for me are nearly that long. 

I just setup 4 P100 GPUs and ran the train_all.py file on 267 patients, batchsize=8, data workers=8
- Each batch for me (batchsize=8) varies. Most batches are 3s but some others are as long as 35s, 45s, 60s.
- Some of the timing bottlenecks don't seem to be reported by time.time() - I think it might be the dataworkers that got deadlocked. 
- Other than that, sometimes the model in/out is really long (60s), sometimes the loading is long (40-60s) and the loss calculation occasionally takes long as well (20s).
- Total time taken for 1 epoch from the start of the training iteration is **7min 14s**.

---- Here are a few of my logs -----
Number of patients training files = 267
- | From start -> setup of model: Time taken = 6.9183571338653564
- | Setup of datasets and train_loader: Time taken = 0.0037386417388916016
0 | Time to load first batch: Time taken = 65.07010841369629
1 | Retrieve data before feeding into model: Time taken = 0.0035915374755859375
2 | Model in -> Model out: Time taken = 18.821953773498535
3 | Calculate loss & update: Time taken = 0.8617253303527832
4 | Zerograd, backprop, optimizer step: Time taken = 14.86911416053772
6 | Time between batches of train_loader: Time taken = 0.00020647048950195312
1 | Retrieve data before feeding into model: Time taken = 0.09489107131958008
2 | Model in -> Model out: Time taken = 0.34704089164733887
3 | Calculate loss & update: Time taken = 0.4055967330932617
4 | Zerograd, backprop, optimizer step: Time taken = 1.6324901580810547
5 | Finished one batch: Time taken = 2.480588912963867
6 | Time between batches of train_loader: Time taken = 6.747245788574219e-05
2019-08-24 05:53:26,200 Iter 2, Epoch 0.0599, Loss 0.9904894
1 | Retrieve data before feeding into model: Time taken = 0.0002377033233642578
2 | Model in -> Model out: Time taken = 0.06069469451904297
3 | Calculate loss & update: Time taken = 0.6516270637512207
4 | Zerograd, backprop, optimizer step: Time taken = 1.6323268413543701
5 | Finished one batch: Time taken = 2.3456132411956787
2019-08-24 05:53:28,546 Iter 3, Epoch 0.0899, Loss 0.9738502
1 | Retrieve data before feeding into model: Time taken = 0.0002281665802001953
2 | Model in -> Model out: Time taken = 59.20406985282898
3 | Calculate loss & update: Time taken = 0.7082192897796631
4 | Zerograd, backprop, optimizer step: Time taken = 1.6912016868591309
5 | Finished one batch: Time taken = 61.634665727615356
2019-08-24 05:57:05,227 Iter 20, Epoch 0.5993, Loss 0.9007623

**Comment by China-LiuXiaopeng** [comment]
It seems that the training process is blocking with multiply workers here and confused me as well. Maybe you could check your torch version (0.4 or 1.0). I have ever met the dataloader issue that my code (not this one) works well under python2+torch0.3 but failed under python3+torch0.3, it is dead in the dataloader (always waiting....), I tackled the issue by updating the torch to 1.0 version.

 The healthy training log should be (batchsize=8, four 1080Ti):
2019-01-19 23:01:23,700 Iter 1, Epoch 0.0316, Loss 0.9967
2019-01-19 23:01:25,539 Iter 2, Epoch 0.0632, Loss 0.9880
2019-01-19 23:01:27,364 Iter 3, Epoch 0.0947, Loss 0.9813
2019-01-19 23:01:29,190 Iter 4, Epoch 0.1263, Loss 0.9693
2019-01-19 23:01:31,016 Iter 5, Epoch 0.1579, Loss 0.9390
2019-01-19 23:01:32,840 Iter 6, Epoch 0.1895, Loss 0.9832
2019-01-19 23:01:34,664 Iter 7, Epoch 0.2211, Loss 0.9250
2019-01-19 23:01:36,546 Iter 8, Epoch 0.2526, Loss 0.9647
2019-01-19 23:01:38,374 Iter 9, Epoch 0.2842, Loss 0.9490
2019-01-19 23:01:40,242 Iter 10, Epoch 0.3158, Loss 0.9510
2019-01-19 23:01:42,139 Iter 11, Epoch 0.3474, Loss 0.8744
2019-01-19 23:01:44,129 Iter 12, Epoch 0.3789, Loss 0.8663
2019-01-19 23:01:45,973 Iter 13, Epoch 0.4105, Loss 0.8082
2019-01-19 23:01:47,851 Iter 14, Epoch 0.4421, Loss 0.8032
2019-01-19 23:01:49,803 Iter 15, Epoch 0.4737, Loss 0.8322
2019-01-19 23:01:51,981 Iter 16, Epoch 0.5053, Loss 0.8499
...
2019-01-19 23:02:55,920 Iter 47, Epoch 1.4842, Loss 0.5505
2019-01-19 23:02:57,926 Iter 48, Epoch 1.5158, Loss 0.5341
2019-01-19 23:03:00,582 Iter 49, Epoch 1.5474, Loss 0.5660
2019-01-19 23:03:02,725 Iter 50, Epoch 1.5789, Loss 0.5569
2019-01-19 23:03:04,738 Iter 51, Epoch 1.6105, Loss 0.6413
2019-01-19 23:03:06,975 Iter 52, Epoch 1.6421, Loss 0.5356
2019-01-19 23:03:09,081 Iter 53, Epoch 1.6737, Loss 0.5404
2019-01-19 23:03:11,155 Iter 54, Epoch 1.7053, Loss 0.6251
...
My final converged loss would fluctuate about 0.32~0.40. By the way,  BraTS 2019 it also works.

**Comment by darrenjkt** [comment]
Hmm that's weird. I have no idea why it's so different. So your current version with no dataloader delays is with python3 and pytorch 1.0? Does yours work with pytorch 1.2? 

I am currently running python 3.7.4 and pytorch 1.2.0.

**Comment by China-LiuXiaopeng** [comment]
I am not sure why it crashed in python 3.7 and pytorch 1.2 just like you described.  I have tried that Pytorch 0.4 and 1.0 with python 3.6 both work OK. Hoping my answer would be helpful to you.

**Comment by darrenjkt** [comment]
I downgraded pytorch 1.0 and it still trains with the same speed but that's okay. I will make do with it. Thank you for your help!

**Comment by darrenjkt** [comment]
Hi I thought I'd follow up on this. 

I've shifted to using 4x2080Ti GPUs and they've gotten training times as you've reported of around 10-11hrs. I think Google Cloud Platform's P100s seem to be quite a bit slower, not sure why. Also the dataworkers don't seem to be stuck anymore.

**Comment by thanhhau097** [comment]
I’ve faced the dataloader delay problem with python 3.7 and pytorch 1.0/1.4. Has anyone solved this problem.
